tools/meteor-services/deploy.js

```diff
@@ -321,6 +323,206 @@ function canonicalizeSite(site) {
+      versionStatusResult.payload &&
+      versionStatusResult.payload.buildStatus) {
+      // Collect all of the statuses from the returned rpc call
+      const {buildStatus, deployStatus, isActive} = versionStatusResult.payload;
+
+      // If the build status has changed, we should report on that change to the user
+      // Allowable values for buildStatus are :
+      // ['missing', 'failed-temporarily', 'failed-permanently', 'success', 'building']
+      //    We will ignore the 'missing' buildStatus, and treat 'failed-temporarily'
+      //    the same as we do 'building', trusting the server to change the status.
+      if(currentStatus.buildStatus !== buildStatus) {
+        // Update our tracking variable for buildStatus
+        currentStatus.buildStatus = buildStatus;
+        // If we have reached a terminal buildStatus, handle it
+        if (terminalBuildStatuses.indexOf(buildStatus) >= 0) {
+          if(buildStatus == 'success') {
+            Console.info('Successfully built app image, initiating deploy');
+            // Reset the polling rounds to poll faster for a change in deployStatus
+            // Basically, we don't necessarily expect deploying to take a long time simply
+            // because building did.
+            pollingRounds = 1;
+          } else {
+            // Any other terminal status is failure, so report on it
+            Console.info(`Your app failed to build, with buildStatus: ${buildStatus}`);
+            // We also want to stop polling, since deploy cannot succeed with a failed build
+            terminal = true;
+          }
+        } else if (buildStatus !== 'building') {
+          // This will report if the buildStatus changes to a non-terminal, unexpected state
+          //    e.g. failed-temporarily
+          Console.info(`Build state changed to nonterminal state: ${buildStatus}`)
+        }
+      } else if(currentStatus.deployStatus !== deployStatus) {
+        // Deploy status is treated similarly to buildStatus above
+        currentStatus.deployStatus = deployStatus;
+        if (terminalDeployStatuses.indexOf(deployStatus) >= 0) {
+          if(deployStatus == 'success') {
+            Console.info('Your app has been deployed successfully.' +
+              ' Waiting for new version to be active...');
+            // Reset the polling rounds to poll faster for a change in isActive
+            pollingRounds = 1;
+          } else {
+            Console.info(`Your app has failed to deploy, with deployStatus: ${deployStatus}`);
+            terminal = true;
+          }
+        } else if(deployStatus !== 'deploying') {
+          Console.info(`Deploy status changed to nonterminal state: ${deployStatus}`)
+        }
+      } else if(currentStatus.isActive !== isActive) {
+        currentStatus.isActive = isActive;
+        // isActive was initialized to false, so it can only be true
+        Console.info('Your deployed version is now active.');
+        terminal = true;
```

Comment:
Rather than flopping `terminal` to `true` here and on lines above (e.g. 470, 488), in order to avoid the conditional on line 506, could we just `return` in these locations and get rid of the `terminal` variable?  If necessary, the `return` status could dictate whether additional recursion is necessary.